namesys/pubsub.go

```diff
@@ -0,0 +1,346 @@
+	if err != nil {
+		return 0, err
+	}
+
+	var entry pb.IpnsEntry
+	err = proto.Unmarshal(dsrec.GetValue(), &entry)
+	if err != nil {
+		return 0, err
+	}
+
+	return entry.GetSequence(), nil
+}
+
+func (p *PubsubPublisher) publishRecord(ctx context.Context, k ci.PrivKey, value path.Path, seqno uint64, eol time.Time, ipnskey string, ID peer.ID) error {
+	entry, err := CreateRoutingEntryData(k, value, seqno, eol)
+	if err != nil {
+		return err
+	}
+
+	data, err := proto.Marshal(entry)
+	if err != nil {
+		return err
+	}
+
+	// the datastore is shared with the routing publisher to properly increment and persist
+	// ipns record sequence numbers; so we need to Record our new entry in the datastore
+	dsrec, err := record.MakePutRecord(k, ipnskey, data, true)
+	if err != nil {
+		return err
+	}
+
+	dsdata, err := proto.Marshal(dsrec)
+	if err != nil {
+		return err
+	}
+
+	p.ds.Put(dshelp.NewKeyFromBinary([]byte(ipnskey)), dsdata)
+
+	// now we publish, but we also need to bootstrap pubsub for our messages to propagate
+	topic := "/ipns/" + ID.Pretty()
+
+	p.mx.Lock()
+	_, ok := p.subs[topic]
+
+	if !ok {
+		p.subs[topic] = struct{}{}
+		p.mx.Unlock()
+
+		bootstrapPubsub(p.ctx, p.cr, p.host, topic)
+	} else {
+		p.mx.Unlock()
+	}
+
+	log.Debugf("PubsubPublish: publish IPNS record for %s", topic)
+	return p.ps.Publish(topic, data)
+}
+
+// Resolve resolves a name through pubsub and default depth limit
+func (r *PubsubResolver) Resolve(ctx context.Context, name string) (value path.Path, err error) {
+	return r.ResolveN(ctx, name, DefaultDepthLimit)
+}
+
+// ResolveN resolves a name through pubsub with the specified depth limit
+func (r *PubsubResolver) ResolveN(ctx context.Context, name string, depth int) (value path.Path, err error) {
+	return resolve(ctx, r, name, depth, "/ipns/")
+}
+
+func (r *PubsubResolver) resolveOnce(ctx context.Context, name string) (value path.Path, err error) {
+	log.Debugf("PubsubResolve: '%s'", name)
+
+	// retrieve the public key once (for verifying messages)
+	xname := strings.TrimPrefix(name, "/ipns/")
+	hash, err := mh.FromB58String(xname)
+	if err != nil {
+		log.Warningf("PubsubResolve: bad input hash: [%s]", xname)
+		return
+	}
+
+	pubk, err := r.pkf.GetPublicKey(ctx, peer.ID(hash))
+	if err != nil {
+		log.Warningf("PubsubResolve: error fetching public key: %s [%s]", err.Error(), xname)
+		return
+	}
+
+	// the mutex is locked both for subscription map manipulation and datastore
+	// manipulation (otherwise race in TTL checking/invalidation)
+	r.mx.Lock()
+	defer r.mx.Unlock()
+
+	// see if we already have a pubsub subscription; if not, subscribe
+	sub, ok := r.subs[name]
+	if !ok {
+		sub, err = r.ps.Subscribe(name)
+		if err != nil {
+			return
+		}
+
+		r.subs[name] = sub
+		go r.handleSubscription(sub, name, pubk)
+		go bootstrapPubsub(r.ctx, r.cr, r.host, name)
+	}
+
+	// resolve to what we may already have in the datastore
+	dsval, err := r.ds.Get(dshelp.NewKeyFromBinary([]byte(name)))
```

Comment:
datastores in the real world will only store and return bytes, we will need to serialize and deserialize